process/block/shardblock.go

```diff
@@ -650,11 +652,24 @@ func (sp *shardProcessor) CommitBlock(
 		log.Debug(errNotCritical.Error())
 	}
 
-	errNotCritical = sp.forkDetector.AddHeader(header, headerHash, process.BHProcessed)
+	finalHeader, errNotCritical := sp.getHighestHdrForOwnShardFromMetachain(header.Round)
+	if errNotCritical != nil {
+		log.Debug(errNotCritical.Error())
+	}
+
+	finalHeaderHash, errNotCritical := core.CalculateHash(sp.marshalizer, sp.hasher, finalHeader)
 	if errNotCritical != nil {
 		log.Debug(errNotCritical.Error())
 	}
 
+	errNotCritical = sp.forkDetector.AddHeader(header, headerHash, process.BHProcessed, finalHeader, finalHeaderHash)
+	if errNotCritical != nil {
+		log.Debug(errNotCritical.Error())
+	}
+
+	hdrsToAttestFinality := uint32(header.Nonce - finalHeader.Nonce)
+	sp.removeNotarizedHdrsBehindFinal(hdrsToAttestFinality)
```

Comment:
it would be cleaner if you send the finalHeader.Nonce than in remove you iterate and remove all headers with nonce smaller that that.

Reply:
Explained above. We can do this, but we should make two separate methods, one for each processor.